homeassistant/components/mqtt.py

```diff
@@ -0,0 +1,164 @@
+"""
+homeassistant.components.mqtt
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+MQTT component, using paho-mqtt. This component needs a MQTT broker like
+Mosquitto or Mosca. The Eclipse Foundation is running a public MQTT server
+at iot.eclipse.org. If you prefer to use that one, keep in mind to adjust
+the topic/client ID and that your messages are public.
+
+Configuration:
+
+To use MQTT you will need to add something like the following to your
+config/configuration.yaml
+
+mqtt:
+  broker: 127.0.0.1
+  port: 1883
+  topic: home-assistant
+  keepalive: 60
+  client_id: home-assistant
+  qos: 0
+  retain: 0
+
+For sending test messages:
+$ mosquitto_pub -h 127.0.0.1 -t home-assistant/switch/1/on -m "Switch is ON"
+For reading the messages:
+$ mosquitto_sub -h 127.0.0.1 -v -t "home-assistant/#"
+"""
+import logging
+
+from homeassistant.helpers import validate_config
+from homeassistant.components.protocol import DOMAIN
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, EVENT_HOMEASSISTANT_STOP)
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "mqtt"
+DEPENDENCIES = []
+MQTT_CLIENT = None
+MQTT_SEND = 'mqtt_send'
+MQTT_RECEIVED = 'mqtt_received'
+
+
+def setup(hass, config):
+    """ Get the MQTT protocol service. """
+
+    if not validate_config(config,
+                           {DOMAIN: ['broker',
+                                     'port',
+                                     'topic',
+                                     'keepalive',
+                                     'client_id']},
+                           _LOGGER):
+        return False
+
+    try:
+        import paho.mqtt.client as mqtt
+
+    except ImportError:
+        _LOGGER.error("Error while importing dependency paho-mqtt.")
+
+        return False
+
+    global MQTT_CLIENT
+    MQTT_CLIENT = MQTT(config[DOMAIN]['broker'],
+                       config[DOMAIN]['port'],
+                       config[DOMAIN]['topic'],
+                       config[DOMAIN]['keepalive'],
+                       config[DOMAIN]['client_id'])
+
+    def stop_mqtt(event):
+        """ Stop MQTT component. """
+        MQTT_CLIENT.stop()
+
+    def start_mqtt(event):
+        """ Launch MQTT component when Home Assistant starts up. """
+        try:
+            MQTT_CLIENT.run()
+
+        except ConnectionRefusedError:
+            _LOGGER.error("Can't connect to the broker. "
```

Comment:
* Can you use `_LOGGER.exception` ? That way it will write the exception to the log too which might aid in debugging.

* Let's put a return after this error so that further code in this method is not executed.